### PR TITLE
Add details for gradle custom plugin.

### PR DESCRIPTION
**Pull request by prudhvigodithi**
Signed-off-by: pgodithi <pgodithi@amazon.com>

Custom plugin's readme

### Description
This PR will add a doc for gradle custom plugins and explains in details on how to use `opensearch.pluginzip` plugin

### Issues Resolved
https://github.com/opensearch-project/opensearch-plugins/issues/143

### Associated issues
https://github.com/opensearch-project/opensearch-build/issues/1916

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prudhvigodithi** [comment]
Hey @dblock can you review this PR?
I will link this doc to the issue's created for each plugin repo as as part of the campaign. 
Example the issue will be something like this https://github.com/prudhvigodithi/opensearch-build/issues/17

File: GRADLE_CUSTOM_PLUGINS.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
+- [Custom Gradle Plugins](#custom-gradle-plugins)
+  - [`opensearch.pluginzip`](#opensearchpluginzip)
```
Comment (gaiksaya) [inline GRADLE_CUSTOM_PLUGINS.md:2]: Make it normal bullet point instead of code (remove quotes)

**Comment by prudhvigodithi** [comment]
Thanks @gaiksaya, added all your suggestions in the latest commits.

**Review by dblock — CHANGES_REQUESTED** [review]
I think this belongs in "Building Plugins with OpenSearch", ie. BUILDING.md.

I'd also want to see examples implemented in https://github.com/opensearch-project/opensearch-plugin-template-java that match the doc.

**Comment by prudhvigodithi** [comment]
Hey @dblock, you suggest to move the data of `GRADLE_CUSTOM_PLUGINS.md`(added via this PR) to `BUILDING.md`?
Also Coming to [opensearch-plugin-template-java](https://github.com/opensearch-project/opensearch-plugin-template-java), this I see for OpenSearch plugins but not for gradle plugins (under buildSrc/), did I miss anything from you previous comment ?

**Comment by prudhvigodithi** [comment]
Hey @dblock  migrated the details to `BUILDING.md`.
PR for [opensearch-plugin-template-java](https://github.com/opensearch-project/opensearch-plugin-template-java/pull/29)
Once this is merged I will add a new PR for opensearch-plugin-template-java to add details about `opensearch.pluginzip`

File: BUILDING.md [hunk 2]
```diff
@@ -162,3 +166,99 @@ publishing {
...
 ```
 Use `./gradlew publishShadowPublicationToStagingRepository` to produce maven artifacts. When building as part of the monolithic distribution customize `scripts/build.sh` to collect maven artifacts. See [job-scheduler#71](https://github.com/opensearch-project/job-scheduler/pull/82/files) for an example.
+
+## Custom Gradle Plugins
+
+Automations, Testing other OpenSearch funtionalities can be acheived using Custom Gradle Plugins, there are several custom gradle plugins developed depending on desired use case. The custom gradle plugins are added in [OpenSearch](https://github.com/opensearch-project/OpenSearch/tree/main/buildSrc) repo.
+
+### opensearch.pluginzip
+
+This plugin will identify the generated zip file distribution which is the ouput of `bundlePlugin` task and publish to local maven repo with right maven coordinates.
```
Comment (dblock) [inline BUILDING.md:176]: will identify -> identifies

and publish -> and publishes

right -> standardized

File: BUILDING.md [hunk 3]
```diff
@@ -162,3 +166,99 @@ publishing {
...
+[Plugin Code](https://github.com/opensearch-project/OpenSearch/tree/main/buildSrc/src/main/java/org/opensearch/gradle/pluginzip), [Plugin Tests](https://github.com/opensearch-project/OpenSearch/tree/main/buildSrc/src/test/java/org/opensearch/gradle/pluginzip), [Plugin META-INF](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/resources/META-INF/gradle-plugins/opensearch.pluginzip.properties)
+
+
+#### Plugin Design
```
Comment (dblock) [inline BUILDING.md:180]: I would remove this title.

File: BUILDING.md [hunk 4]
```diff
@@ -162,3 +166,99 @@ publishing {
...
+* This plugin requires build-tools framework as dependancy.
+* The maven coordinates groupID is fixed as `org.openserach.plugin`, `version` and `artifcatID` will be inferred from gradle project properties.
+* User can also pass custom POM extensions, that will add desired xml to the zip maven POM file geneaterd during runtime.
+* Once the plugin is added to the `build.gradle` as `apply plugin: 'opensearch.pluginzip'`, this will add a new custom publish task `publishPluginZipPublicationToZipStagingRepository`, this task will publish the zip distribution to the maven local (file system).
+* The maven local artifacts can then be published to maven central/nexus.
+* The plugin will not add jars generated by tasks `sourcesJar` and `javadocJar`, this is done in purpose to exclude `jars` for `zip` publications.
+
+#### Plugin Usage
+
+* The Key requirement for this plugin to work is that the zip should be generated by `bundlePlugin` task that comes from existing [PublishPlugin](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/java/org/opensearch/gradle/PublishPlugin.java).
```
Comment (dblock) [inline BUILDING.md:192]: Make the steps numbered, 1.2.3

**Review by dblock — CHANGES_REQUESTED** [review]
I mostly have nitpicks, the doc is good!

File: BUILDING.md (hunk 3 above)
Comment (prudhvigodithi) [reply BUILDING.md:180]: Wont it be easy to check the plugin code location, if a user wants to know more details about this ?

File: BUILDING.md [hunk 5]
```diff
@@ -162,3 +166,99 @@ publishing {
...
+
+    Note: If gradle project is executed by specific task, then its not required to add ```startParameter.excludedTaskNames=["validatePluginZipPom"]```, only required when ran `./gradlew build`. 
+
+* Exclude the following tasks in `settings.gradle` file, if the build script exists for the plugin and has the tasks `publishToMavenLocal` and `publishAllPublicationsToStagingRepository`, these will also include the tasks `publishPluginZipPublicationToMavenLocal` and `publishPluginZipPublicationToStagingRepository` which is not required to be called with this plugin.
+To exclude add the following in `settings.gradle` file ```startParameter.excludedTaskNames=["publishPluginZipPublicationToMavenLocal", "publishPluginZipPublicationToStagingRepository"]```
+
+    Note: If there is a managed `build.sh` file and do not have any publish tasks, then its not required to exlcude these tasks, only required if it is calling publish tasks that targets ALL repos and ALL publications.
+
+* Quick Example:
```
[the remark this answers is not in the document]
Comment (prudhvigodithi) [reply BUILDING.md:259]: @dblock the plugins have not yet started to use to this gradle plugin, so we dont yet have a PR, once this document is merged and I have a PR as an example, then I can update the doc back again.

File: BUILDING.md [hunk 6]
```diff
@@ -6,6 +6,10 @@
   - [Always Build SNAPSHOT Artifacts](#always-build-snapshot-artifacts)
   - [Consume Maven Artifacts in Order](#consume-maven-artifacts-in-order)
   - [Include Checksums in Maven Publications](#include-checksums-in-maven-publications)
+  - [Custom Gradle Plugins](#custom-gradle-plugins)
+    - [opensearch.pluginzip](#opensearchpluginzip)
+        - [Plugin Design](#plugin-design)
```
[the remark this answers is not in the document]
Comment (prudhvigodithi) [reply BUILDING.md:11]: would it not be easy for a user to read sub-sections rather than a bit fat single section about the plugin? :)

**Comment by bbarani** [comment]
> I mostly have nitpicks, the doc is good!

@dblock Can we go ahead and merge this PR and work on the nitpicks later? I would like the teams to start publishing the plugin Zips starting 2.0 GA version.

**Comment by prudhvigodithi** [comment]
Hey @dblock I have pushed some suggested NIT's, please review.
Thanks

File: BUILDING.md (hunk 4 above)
Comment (dblock) [reply BUILDING.md]: It's not an enumeration, I'd just write the text as is.

File: BUILDING.md (hunk 3 above)
Comment (dblock) [reply BUILDING.md:180]: The documentation for the plugin belongs with the plugin in its code. You can add these things as comments in the plugin itself. This doc is just about how to use it.

File: BUILDING.md [hunk 7]
```diff
@@ -162,3 +166,101 @@ publishing {
...
+* `opensearch.pluginzip` is java based code published into build-tools.
+* The maven coordinates groupID is fixed as `org.openserach.plugin`, `version` and `artifcatID` will be inferred from gradle project properties.
+* User can also pass custom POM extensions, that will add desired xml to the zip maven POM file geneaterd during runtime.
+* Once the plugin is added to the `build.gradle` as `apply plugin: 'opensearch.pluginzip'`, this will add a new custom publish task `publishPluginZipPublicationToZipStagingRepository`, this task will publish the zip distribution to the maven local (file system).
+* The maven local artifacts can then be published to maven central/nexus.
+* The plugin will not add jars generated by tasks `sourcesJar` and `javadocJar`, this is done in purpose to exclude `jars` for `zip` publications.
+
+#### Plugin Usage
+
+1. The Key requirement for this plugin to work is that the zip should be generated by `bundlePlugin` task that comes from existing [PublishPlugin](https://github.com/opensearch-project/OpenSearch/blob/main/buildSrc/src/main/java/org/opensearch/gradle/PublishPlugin.java).
```
Comment (dblock) [inline BUILDING.md:191]: Nit: try to shorten things, e.g. "The Key requirement for this plugin to work is that the zip should be generated by `bundlePlugin` task" is the same as "Ensure that your project uses bundlePlugin".

**Review by dblock — APPROVED** [review]
This looks good! I would just move the whole plugin design thing out. But also you can iterate on this later.

File: BUILDING.md (hunk 7 above)
Comment (prudhvigodithi) [reply BUILDING.md]: Fixed this @dblock

**Review by dblock — APPROVED** [review]

**Review by bbarani — APPROVED** [review]

**Pull request merged**